paddle/fluid/framework/details/all_reduce_op_handle.cc

```diff
@@ -107,14 +107,25 @@ void AllReduceOpHandle::RunImpl() {
           PADDLE_ENFORCE(platform::dynload::ncclAllReduce(
               buffer, buffer, numel, static_cast<ncclDataType_t>(dtype),
               ncclSum, comm, stream));
+          if (!nccl_ctxs_->need_group_call_) cudaStreamSynchronize(stream);
         });
       }
+
       this->RunAndRecordEvent([&] {
-        platform::NCCLGroupGuard guard;
-        for (auto &call : all_reduce_calls) {
-          call();
+        // TODO(Yancey1989): need allreduce operator to avoid this flag
+        if (nccl_ctxs_->need_group_call_) {
+          platform::NCCLGroupGuard guard;
+          for (auto &call : all_reduce_calls) {
+            call();
+          }
+        } else {
+          // only used in executor_type == ParallalGraph, one thread one GPU
+          // TODO(Yancey1989): use allreduce operator to avoid this tricky.
+          PADDLE_ENFORCE(all_reduce_calls.size() == 1UL);
+          all_reduce_calls[0]();
```

Comment:
It seems that it is easy to produce deadlock as described in https://arxiv.org/pdf/1706.02677.pdf

Reply:
Sure, the same problem with distributed training with [NCCL2 collective mode](http://paddlepaddle.org/documentation/docs/zh/1.2/user_guides/howto/training/cluster_howto.html#permalink-8--nccl2-), and we need to fix the order of the all-reduce operators, @gongweibao has a PR to do that https://github.com/PaddlePaddle/Paddle/pull/14586 , I will do more test with that feature after this PR merged.